core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -0,0 +1,432 @@
+            }
+
+            final long oldRequested = requested;
+            if (oldRequested == Long.MAX_VALUE) {
+                return;
+            }
+            if (n == Long.MAX_VALUE) {
+                requested = Long.MAX_VALUE;
+            } else {
+                requested = LongMath.saturatedAdd(oldRequested, n);
+            }
+
+            if (oldRequested > 0) {
+                // SurroundingSubscriber is publishing data.
+                // New requests will be handled by 'publishDownstream(item)'.
+                return;
+            }
+
+            publish();
+        }
+
+        private void publish() {
+            if (state == State.DONE || requested <= 0 && upstreamRequested <= 0) {
+                return;
+            }
+
+            switch (state) {
+                case REQUIRE_HEAD: {
+                    sendHead();
+                    break;
+                }
+                case REQUIRE_BODY: {
+                    if (!subscribed) {
+                        subscribed = true;
+                        publisher.subscribe(this, executor, options);
+                        return;
+                    }
+                    if (upstreamRequested > 0) {
+                        return;
+                    }
+                    final Subscription upstream = this.upstream;
+                    if (upstream != null) {
+                       requestUpstream(upstream);
+                    }
+                    break;
+                }
+                case REQUIRE_TAIL: {
+                    sendTail();
+                    break;
+                }
+                case REQUIRE_COMPLETE: {
+                    close0(null);
+                    break;
+                }
+            }
+        }
+
+        private void sendHead() {
+            setState(State.REQUIRE_HEAD, State.REQUIRE_BODY);
+            assert head != null;
+            publishDownstream(head, true);
+        }
+
+        private void sendTail() {
+            setState(State.REQUIRE_TAIL, State.REQUIRE_COMPLETE);
+            if (tail != null) {
+                downstream.onNext(tail);
+            }
+            close0(null);
+        }
+
+        private void requestUpstream(Subscription subscription) {
+            if (requested <= 0) {
+                return;
+            }
+            assert upstreamRequested == 0;
+            upstreamRequested = requested;
+            if (requested < Long.MAX_VALUE) {
+                requested = 0;
+            }
+            subscription.request(upstreamRequested);
+        }
+
+        private void publishDownstream(T item, boolean head) {
+            requireNonNull(item, "item");
+            if (state == State.DONE) {
+                PooledObjects.close(item);
+                return;
+            }
+            downstream.onNext(item);
+
+            if (head) {
+                if (requested < Long.MAX_VALUE) {
+                    requested--;
+                }
```

Comment:
Can you check if the following scenario passes?

```
final HttpResponse response = HttpResponse.of(ResponseHeaders.of(200), Mono.empty());
final AtomicBoolean complete = new AtomicBoolean();
response.subscribe(new Subscriber<HttpObject>() {
    @Override
    public void onSubscribe(Subscription s) {
        s.request(1);
    }

    @Override
    public void onNext(HttpObject httpObject) {

    }

    @Override
    public void onError(Throwable t) {
        complete.set(true);
    }

    @Override
    public void onComplete() {
        complete.set(true);
    }
});
await().until(complete::get);
```

Reply:
> On publisher with 1 element, I think onComplete should be delivered after request(1) twice.

Just to make sure I understood correctly, are you suggesting that in this case we need to request `2` to receive `1` element and complete?

I doubt this is how most implementations works 😅    (actually, even `SurroundingPublisher` itself seems to be inconsistent since most times the stream is completed without an extra request)

Reply:
> On publisher with 1 element, I think onComplete should be delivered after request(1) twice.

Just to make sure I understood correctly, are you suggesting that in this case we need to request 2 to receive 1 element and complete?

I doubt this is how most implementations works 😅 (actually, even SurroundingPublisher itself seems to be inconsistent since most times the stream is completed without an extra request)

Reply:
oh.. after some thinking, I think we should preemptively subscribe upstream on `sendHead()` for consistent implementation. (ensure elements: 1 -> request(1) -> `onNext&onComplete`)

I checked that TCK test passed on above case~! I'll update PR! (I'll also add related unit-test too 🙇)

Reply:
@jrhee17 I updated PR based on your suggestion and add more unit-tests!

Now your scenario ([comment](https://github.com/line/armeria/pull/4727#discussion_r1235142493)) pass test well 👍

Reply:
> What do you think of preemptively subscribing to upstream once head is published?

The current approach is reasonable. However, I have one thing to note here. As `Subscriber` does not know the number of elements in the publisher preemptively, I don't think we need to generally consider https://github.com/line/armeria/pull/4727/files#r1235142493 case.